src/storage/in_memory.rs

```diff
@@ -10,36 +10,35 @@ type Accounts = HashMap<Address, EvmAccount, BuildAddressHasher>;
 
 /// A storage that stores chain data in memory.
 #[derive(Debug, Default, Clone)]
-pub struct InMemoryStorage {
+pub struct InMemoryStorage<'a> {
     accounts: Accounts,
-    bytecodes: AHashMap<B256, EvmCode>,
+    bytecodes: Option<&'a AHashMap<B256, EvmCode>>,
     block_hashes: AHashMap<u64, B256>,
 }
 
-impl InMemoryStorage {
+impl<'a> InMemoryStorage<'a> {
     /// Construct a new [InMemoryStorage]
-    // TODO: Take in [bytecodes] instead of reading duplicates from
-    // [accounts].
     pub fn new(
         accounts: impl IntoIterator<Item = (Address, EvmAccount)>,
+        bytecodes: Option<&'a AHashMap<B256, EvmCode>>,
         block_hashes: impl IntoIterator<Item = (u64, B256)>,
     ) -> Self {
         let accounts: Accounts = accounts.into_iter().collect();
-        let mut bytecodes = AHashMap::default();
-        for (_, account) in accounts.iter() {
-            if let (Some(code_hash), Some(code)) = (account.code_hash, &account.code) {
-                bytecodes.entry(code_hash).or_insert_with(|| code.clone());
-            }
+        let block_hashes: AHashMap<u64, B256> = block_hashes.into_iter().collect();
+
+        for account in accounts.values() {
+            assert!(account.code.is_none());
```

Comment:
This is stricter than needed since these `code`s are unused, their actual values don't affect correctness. We should just inline the `collect`s and return an `InMemoryStorage` directly.

Reply:
ok

Reply:
Fixed.

Keep in mind that this `assert!` statement has shown its effectiveness in the recent CI failure: https://github.com/risechain/pevm/actions/runs/10263383045/job/28395093118?pr=245

Without this `assert!`, the error log will be unmeaningful:
```
---- ethereum_state_tests stdout ----
thread '<unnamed>' panicked at tests/ethereum/main.rs:219:21:
assertion `left == right` failed: Mismatched state root for "tests/ethereum/tests/GeneralStateTests/stRandom2/randomStatetest534.json"
  left: 0x0dc060d876225462f5269e64bd3bcea7b687d760181636f9d6b1805657b1121b
 right: 0x0746899e8148565c59dd0b63c61acafc2b1ac75aa977c3a620b2b3a9c0667a5f
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at tests/ethereum/main.rs:219:21:
assertion `left == right` failed: Mismatched state root for "tests/ethereum/tests/GeneralStateTests/stRevertTest/RevertInDelegateCall.json"
  left: 0xe6960d03cae79440ffbf8b3deeeebc75da0b69b812aed7170c05b5949816aea7
 right: 0xd30a9c46d689c5fead9ce54e1b2f441bde1ed776daf747d9de01df716bbbbe2c
```

Reply:
Pretty nice but not worth it. Also, if we aim for helpfulness then we should error when there are codes in accounts but `bytecodes` is `None`. Having codes in both seem like valid usage and wasn't the root cause of the failed test anyway.